### PR TITLE
Updates for Release 2

### LINKED ISSUE #8: Sudoku Tester
This function will make sure whether the generated sudokus are following the regulations.

### PR DESCRIPTION
**Pull request by Prasanthkota165**
Any corrections for Release 1 and new features of Release 2 are to be added.

### CODE REVIEW

**Review by Prasanthkota165 — COMMENTED** [review]
Added complete randomness for rows,columns swapping and row,column groups swapping

**Review by Mathieufalardeau3 — COMMENTED** [review]
Where is the sudoku_tester function? The whole point of release 2 was to include the testing_function into the program.

File: src/sudoku_base.c [hunk 1]
```diff
@@ -29,22 +29,31 @@ int main(void){
...
+        /* This function will generate the sudoku template */
+        generate_sudoku_template(sudoku_template);
+
+        /* Rearranging the sudoku template to generate different combinations 
+        of sudoku */ 
+        rearrange_cols_rows(sudoku_template, rearranged_sudoku);
+        
+        /*Tests the rearranged sudoku for sudoku rules*/
+        tester=test_sudoku_rules(rearranged_sudoku);
+
```
Comment (Mathieufalardeau3) [inline src/sudoku_base.c]: We will add a loop here on Tuesday. No need to change anything.

File: src/test_sudoku.c [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            k=k+3;
+            l=l+3;
+            sumplus=sum;
+            while(!isEmpty() || sumplus!=45) {/*checks whether the array is empty or not. If not it empties the array for furthur operation.*/
+                flag=true;  
+                sumplus=45;        
+                struct node *temp = deleteFirst();
+            }  
+            sum=0;
+    	    insertFirst(1);
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c]: The insertFirst() function is not properly spaced

File: src/test_sudoku.c [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+#include <stdbool.h>
+#include "../include/sudoku.h"
+
+struct node {
+    int data;
+    struct node *next;
+};
+
+struct node *head = NULL;
+struct node *current = NULL;
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c]: The teacher mentioned that we should not use Global variables.

File: include/sudoku.h [hunk 4]
```diff
@@ -49,11 +49,25 @@ void generate_sudoku_template(int sudoku_temp[ROWS][COLS]);
...
 */
-void rearrange_cols_rows(int sudoku_temp[ROWS][COLS], 
-                            int rearranged_sudoku_temp[ROWS][COLS]);
-
-_Bool test_sudoku_rules(int test_sudoku[ROWS][COLS]);
+void rearrange_cols_rows(int [ROWS][COLS], int [ROWS][COLS]);
 
+/**
+ * The function test_sudoku_rules checks the rearranged sudoku (which is generated in
+ * the function rearrange_cols_rows) for the sudoku rules. It examines the rows, coloums
```
Comment (Mathieufalardeau3) [inline include/sudoku.h:56]: "...It examines the rows, **columns...**"

File: src/test_sudoku.c [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+/*
+ *This checks the 3X3 blocks for repeated values.
+ */
+_Bool test_3X3(int rearranged_sudoku[][9]) {
+    int z[9][9];  
+    int x=0,y=2,sum=0,k=0,l=2,a,sumplus=0;
+    int flag=0;
+    while(x<=8) {
+        while(k<=8) {
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c:98]: I see that your code checks for a 3x3 block, but what about cols and rows? I don't see how this function can verify if two identical digits are not in the same row or col?

File: src/test_sudoku.c [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+/**
+ * Insert link at the first location.
+ */
+void insertFirst(int data){
+    struct node *link = (struct node*) malloc(sizeof(struct node));/*create a link*/
+    link->data = data;
+    link->next = head;/*point it to old first node*/
+    head = link;/*point first to new first node*/
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c]: Why are you using linked list? Linked list are used when you need to rearrange the data in a certain order. The sudoku_tester() function should be about comparing elements to determine if two digits repeat themselves. I don't see the use of linked list when you can easily just create a copy of 2D array and simply create functions to verify the rows, cols and 3x3 blocks.

File: src/test_sudoku.c [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        previous->next = current->next;/*bypass the current link*/
+     }    
+    return current;
+}
+
+/*
+ *This checks the 3X3 blocks for repeated values.
+ */
+_Bool test_3X3(int rearranged_sudoku[][9]) {
+    int z[9][9];
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c]: This variable is not used in the function.

File: src/test_sudoku.c [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    return flag;
+}
+
+/**
+ * This function checks the rows, columns and 3X3 matrix for sudoku rules.
+ */
+_Bool test_sudoku_rules(int rearranged_sudoku[][9]) {  
+    test_3X3(rearranged_sudoku);
+    if(test_3X3(rearranged_sudoku)){
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c]: Why are you calling the function test_3x3 twice? The if() statement already calls the test 3x3 function. It does not need to be called before that. It is redundant.

File: src/test_sudoku.c [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            sum=0;
+    	    insertFirst(1);
+            insertFirst(2);
+            insertFirst(3);
+            insertFirst(4);
+            insertFirst(5);
+            insertFirst(6);
+            insertFirst(7);
+            insertFirst(8);
+            insertFirst(9);
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c]: This entire portion is useless. It is not explain anywhere what this does. What I can see is that you are adding the numbers 1 through 9 into a linked list and then deleting them when you are doing your if() statement at the end of the block check. I will be frank. This verifies nothing. It does not verify anything in the rearanged_sudoku array and uses heap memory for nothing. This needs to go.

File: test/src/test_sudoku_test.c [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ * Author   : Rishith Rao - rishithrao@cmail.carleton.ca
+ *            Prasanth Kota - prasanthkota@cmail.carleton.ca
+ *
+ * Date     : 13 June, 2020.
+ */
+
+
+void main() {
+    _Bool a;
+    a=test_sudoku_rules(rearranged_sudoku);
```
Comment (Mathieufalardeau3) [inline test/src/test_sudoku_test.c:33]: Where is the rearranged_sudoku variable is being initialized?

**Review by Mathieufalardeau3 — CHANGES_REQUESTED** [review]
Franky, test_sudoku.c is overly complicated and the use of linked list and global variables is a poor design choice. 
Remember that the Project instructions specified that global variables were not to be used in the project.
 
Guys, this is not good. For the sudoku_solver function, I had to create parts of a test_sudoku function and it was half the size of what you wrote for just checking a 3x3 block.

File: src/test_sudoku.c (hunk 2 above)
Comment (Prasanthkota165) [reply src/test_sudoku.c]: Agreed and modified.

File: src/test_sudoku.c (hunk 5 above)
Comment (Prasanthkota165) [reply src/test_sudoku.c:98]: Tasks are divided between me and Rishith. Rows/Columns will be done by me and Rishith does the rest.

File: src/test_sudoku.c (hunk 6 above)
Comment (Prasanthkota165) [reply src/test_sudoku.c]: There is a use of linked lists here. We are copying the values from 1 to 9 to the linked lists and checking the elements of rows/columns/3x3 have values from 1 to 9 or not. So, the linked list will be left with the values that has to be appeared in the place of the repeated values in the row/column/3x3 blocks. This is also modified to simple function for better understanding for the other developers.

File: src/test_sudoku.c (hunk 9 above)
Comment (Prasanthkota165) [reply src/test_sudoku.c]: There is a purpose to add the elements to linkedlists. This is also modified for better understanding for the other developers.

**Comment by Prasanthkota165** [comment]
Just wait for the rishith to push the test function for 3x3 block and the other developers can do the pull and test them accordingly

File: src/test_sudoku.c [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+#include <stdbool.h>
+#include "../include/sudoku.h"
+
+/**
+ * This function checks the rows, columns and 3X3 matrix for sudoku rules.
+ */
+
+_Bool test_sudoku_rules(int rearranged_sudoku[][9]){  
+  int x,y,z; 
+  int a[9]; /* Initialize array to copy the values 1 to 9 */
```
Comment (Mathieufalardeau3) [inline src/test_sudoku.c:26]: Missing the check box function. I assume it is coming soon.

**Comment by Prasanthkota165** [comment]
Fixed everything. You guys can pull and test them.

**Review by rishithrao57 — APPROVED** [review]
Everything looks good. We can go for the release.

**Review by Mathieufalardeau3 — APPROVED** [review]

**Review by Mathieufalardeau3 — APPROVED** [review]
Good for release 2

**Review by gabriels2018 — APPROVED** [review]
I approve the changes

**Pull request merged**